pkg/daemon/cleanup/cleanup.go

```diff
@@ -194,6 +196,33 @@ func removeServices(svcMgr manager.Manager, services []servicescm.Service) error
 			servicesRemoved = append(servicesRemoved, service.Name)
 		}
 	}
+
+	if bestEffort {
+		// Remove all services that were created through WICD, even if they are no longer in the services ConfigMap spec
+		existingSvcs, err := svcMgr.GetServices()
+		if err != nil {
+			return fmt.Errorf("could not determine existing Windows services: %w", err)
+		}
+		for name := range existingSvcs {
+			winSvcObj, err := svcMgr.OpenService(name)
+			if err != nil {
+				return err
+			}
+			config, err := winSvcObj.Config()
+			if err != nil {
+				return err
+			}
+			if strings.Contains(config.Description, windows.ManagedTag) {
+				if err := svcMgr.DeleteService(name); err != nil {
```

Comment:
Will the priority order of the Windows services be a dependency issue given `existingSvcs` has no order?

Reply:
This is a good point and something I went back and forth about — I chose to not look into dependencies in best effort cleanup because of a particular scenario that could enable a DDOS:
- user sneaks in service X into configmap spec
- Service X depends on a crucial system service like system32
- WICD goes to delete service X
- Traverses dependency tree and first stops system32…

Reply:
I agree but there is a caveat here. We are also using the best effort bool to clean the services from the latest config map if the version config map is not found. Wouldn't we want to delete the services by dependency for that?

Reply:
> services from the latest config map

These are passed in to removeServices and so will be deleted in proper dependency order. The bestEffort block will only affect lingering OpenShift managed services that haven't been passed into the func (i.e. services in neither latest CM nor version CM)

Reply:
https://github.com/openshift/windows-machine-config-operator/pull/2408/files#r1747421558

I'm not sure if this logic is correct.
Why should we stop system32 if Service X depends on it?
We should have no problem stopping service X. The issue would be if system32 depended on service X. In that case we would have to stop system32 before stopping service x.

But in order for that to be the case, the person doing this would need permissions to change the system32 service in the first place.

Reply:
@sebsoto was correct. As a bonus, we realized DeleteService() does a "safe delete" under the hood meaning we already implemented the dependency tree traversal logic behind the scenes :)